### ISSUE TITLE
GPSampler's suggestion fails when used inside `torch.no_grad()` context manager

### ISSUE THREAD

**Issue by kAIto47802** [optuna/optuna#5663]
### Expected behavior

Currently, GPSampler's `sample_relative` fails when used inside `torch.no_grad()` context manager, resulting the default initial kernel params to be used.
Specifically, the `torch.no_grad()` context manager, when used outside of Optuna, affects the PyTorch gradient calculations within GPSampler and raises a error, causing `sample_relative` to be fail.

The expected behavior is:

- GPSampler work correctly even when  the `torch.no_grad()` is active.

### Environment

- Optuna version: 4.0.0
- Python version: 3.11.0
- OS: MacOS, Ubuntu
- torch: 2.4.0

### Error messages, stack traces, or logs

```shell
[I 2024-09-03 07:44:07,293] A new study created in memory with name: no-name-2ddc8765-755f-4bca-8a7a-fd0f3249e467
[I 2024-09-03 07:44:07,322] Trial 0 finished with value: 0.32613587379455566 and parameters: {'n_hidden': 3}. Best is trial 0 with value: 0.32613587379455566.
[W 2024-09-03 07:44:07,560] The optimization of kernel_params failed: element 0 of tensors does not require grad and does not have a grad_fn
The default initial kernel params will be used instead.
[I 2024-09-03 07:44:07,801] Trial 1 finished with value: -0.08969029039144516 and parameters: {'n_hidden': 10}. Best is trial 1 with value: -0.08969029039144516.
```

### Steps to reproduce

The following code yields the error above:

```python
import optuna
import torch
from torch import nn

class Net(nn.Module):
   def __init__(self, n_input, n_output, n_hidden):
       super(Net, self).__init__()
       self.fc1 = nn.Linear(n_input, n_hidden)
       self.fc2 = nn.Linear(n_hidden, n_output)

   def forward(self, x):
       x = torch.relu(self.fc1(x))
       x = self.fc2(x)
       return x

def objective(trial: optuna.Trial) -> float:
   with torch.no_grad():
       n_input = 10
       n_output = 1
       n_hidden = trial.suggest_int('n_hidden', 1, 10)
       net = Net(n_input, n_output, n_hidden)
       x = torch.randn(1, n_input)
       y = net(x)
   return y.item()

sampler = optuna.samplers.GPSampler(n_startup_trials=1)
study = optuna.create_study(sampler=sampler)
study.optimize(objective, n_trials=2)
```

**Comment by kAIto47802** [comment]
I will work on this issue.

**Comment by kAIto47802** [comment]
The cause of the issue is as follows:

https://github.com/optuna/optuna/blob/558f215e5644226fb1a55f09ab4b6ab3d28b57f9/optuna/_gp/gp.py#L175-L194

In the `loss_func` within `_fit_kernel_params` function, 

https://github.com/optuna/optuna/blob/558f215e5644226fb1a55f09ab4b6ab3d28b57f9/optuna/_gp/gp.py#L177

is set. 

However, inside the `torch.no_grad()` context, operations like `torch.exp(raw_params_tensor[:n_params])` do not result in the output having `requires_grad=True` and the loss

https://github.com/optuna/optuna/blob/558f215e5644226fb1a55f09ab4b6ab3d28b57f9/optuna/_gp/gp.py#L187-L189

dose not have the gradient information.

This leads to a failure in performing backpropagation at

https://github.com/optuna/optuna/blob/558f215e5644226fb1a55f09ab4b6ab3d28b57f9/optuna/_gp/gp.py#L190

.

**Comment by kAIto47802** [comment]
One possible solution of this is to set grad enabled with in the `loss_func`:

```py
def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
    raw_params_tensor = torch.from_numpy(raw_params)
    raw_params_tensor.requires_grad_(True)
    with torch.enable_grad():
        params = KernelParamsTensor(
            inverse_squared_lengthscales=torch.exp(raw_params_tensor[:n_params]),
            kernel_scale=torch.exp(raw_params_tensor[n_params]),
            noise_var=(
                torch.tensor(minimum_noise, dtype=torch.float64)
                if deterministic_objective
                else torch.exp(raw_params_tensor[n_params + 1]) + minimum_noise
            ),
        )
        loss = -marginal_log_likelihood(
            torch.from_numpy(X), torch.from_numpy(Y), torch.from_numpy(is_categorical), params
        ) - log_prior(params)
        loss.backward()  # type: ignore
        # scipy.minimize requires all the gradients to be zero for termination.
        raw_noise_var_grad = raw_params_tensor.grad[n_params + 1]  # type: ignore
        assert not deterministic_objective or raw_noise_var_grad == 0
    return loss.item(), raw_params_tensor.grad.detach().numpy()
```

> [!NOTE]
> The context manager [`torch.enable_grad()`](https://pytorch.org/docs/stable/_modules/torch/autograd/grad_mode.html#enable_grad) preserves whether the mode was originally set to compute gradient information, and when `__exit__()` is called, it restores the original state.

**Comment by kAIto47802** [comment]
I will make PR with unit test with that change.
